### PR TITLE
[BUG FIX] Fix buggy hibernation on GPU and correctly report resting force.

### LINKED ISSUE #3293: [Bug]: NaN when enabling hibernation
### Bug Description

To reproduce, run the following script:

```
#!/usr/bin/env python3
"""Minimal reproduction of a rigid-hibernation numerical instability.

Scene: the `table_bussing` benchmark scene -- a fixed 18-DOF bimanual robot posed over a work table with a 4x4 grid
of dining clutter (mug/cup/apple/donut, CoACD-decomposed) dropped to settle on the tabletop. Contact-rich mesh
collision plus a high-DOF articulated system held by PD control.

Bug: with `use_hibernation=True`, the constraint solve diverges to NaN and Genesis raises

    genesis.GenesisException: Invalid constraint forces causing 'nan'. Please decrease Rigid simulation timestep.

The ONLY change from a healthy run is `USE_HIBERNATION`; set it False and the identical scene runs stably.

Reliability: the clutter is periodically re-dropped per env (decorrelated resets, as in the `table_bussing_forever`
benchmark) so the scene never fully settles and the contact-rich transient that trips the solver keeps recurring.
With resets on, this reproduces on every run (GPU, N_ENVS=50, genesis main). Set GS_REPRO_RESET_STEPS=0 to disable
resets and see the flakier static case (~2/3 of runs).

How to reproduce
----------------
Requires a CUDA GPU and a Genesis env with genesis (main) + quadrants + torch + huggingface_hub installed (i.e. the
standard Genesis container's /venv). The script hardcodes backend=gs.gpu and pulls its assets from the pinned HF
dataset, so no manual asset setup or benchmark harness is needed -- it builds the scene itself and runs on plain
genesis main. Inside such an env just run:

    python repro_hibernate_nan.py

Expected: it prints the genesis/quadrants versions, then crashes within ~35 steps with

    *** REPRODUCED at step 35: Invalid constraint forces causing 'nan'. Please decrease Rigid simulation timestep.

and exits nonzero. Knobs:
  * GS_REPRO_RESET_STEPS=0 python repro_hibernate_nan.py  -> static variant (flakier; NaNs later, ~step 50-90)
  * edit USE_HIBERNATION = False (below)                  -> the identical scene runs stably to the end

From scratch on the cluster (how this was verified): allocate a GPU node, start the Genesis container (image from
genesis's .github/genesis_image_ver), build /venv on origin/main, then run:

    cd <genesis-checkout-at-main> && bash .github/workflows/scripts/production_build.sh   # installs genesis+quadrants
    source /venv/bin/activate
    python bugs/repro_hibernate_nan.py

The first run spends ~2-3 min on CoACD decomposition + kernel compile, then the NaN fires almost immediately.
Verified on genesis main (v1.3.3) + quadrants 1.3.0, field dtype: reset mode 5/5 (NaN at step 35-36), static 3/3
(steps 48-89); with use_hibernation=False the same scene is stable (~335 fps).
"""

import os

# Match the condition the bug was observed under (the benchmark runs the 'field' dtype). Must be set before the
# genesis import. Flip to "1" to also check the ndarray backend.
os.environ.setdefault("GS_ENABLE_NDARRAY", "0")

import numpy as np
import torch
from huggingface_hub import snapshot_download

import genesis as gs

# --- the bug switch -------------------------------------------------------------------------------------------
USE_HIBERNATION = True  # set False -> the identical scene is numerically stable
N_ENVS = 50
MAX_STEPS = 600  # ~20 s of sim at dt=1/30; the NaN hits well before this when resets are on
# Per-env decorrelated re-drop interval, in steps (2 s at dt=1/30). 0 => never reset (static, flakier repro).
RESET_STEPS_MAX = int(os.environ.get("GS_REPRO_RESET_STEPS", "60"))
# --------------------------------------------------------------------------------------------------------------

# Same pinned revision as tests/utils/assets.py::get_hf_dataset, inlined so this script is self-contained.
HF_ASSETS_REVISION = "990a727788f11e34ad006c69bf769303b20cb11c"


def get_asset(pattern):
    return snapshot_download(
        repo_type="dataset",
        repo_id="Genesis-Intelligence/assets",
        revision=HF_ASSETS_REVISION,
        allow_patterns=pattern,
        max_workers=1,
    )


def main():
    gs.init(backend=gs.gpu)

    STEP_DT = 1.0 / 30.0
    TABLE_TOP_Z = 0.54  # work_table.glb top face sits ~0.54 above the mesh origin at default scale

    scene = gs.Scene(
        sim_options=gs.options.SimOptions(dt=STEP_DT, substeps=10),
        rigid_options=gs.options.RigidOptions(
            noslip_iterations=5,
            max_collision_pairs=256,
            max_contacts=1024,
            use_hibernation=USE_HIBERNATION,
        ),
        show_viewer=False,
        show_FPS=False,
    )

    scene.add_entity(gs.morphs.Plane())

    table_path = get_asset("work_table.glb")
    scene.add_entity(
        gs.morphs.Mesh(file=f"{table_path}/work_table.glb", pos=(0.5, 0.0, 0.0), fixed=True),
        vis_mode="collision",
    )

    robot_path = get_asset("dual_arms_primitives.urdf")
    robot = scene.add_entity(
        gs.morphs.URDF(file=f"{robot_path}/dual_arms_primitives.urdf", pos=(0.0, 0.0, TABLE_TOP_Z + 0.5), fixed=True),
    )

    assets = (("mug_1", "output.xml"), ("cup_2", "model.xml"), ("apple_15", "model.xml"), ("donut_0", "output.xml"))
    asset_files = {name: f"{get_asset(f'{name}/*')}/{name}/{xml}" for name, xml in assets}
    for i in range(16):
        gx, gy = i % 4, i // 4
        name = assets[(gx + gy) % len(assets)][0]
        scene.add_entity(
            gs.morphs.MJCF(
                file=asset_files[name],
                pos=(0.30 + 0.11 * gx, -0.17 + 0.11 * gy, TABLE_TOP_Z + 0.06 + 0.02 * i),
                euler=(90.0, 0.0, 0.0),
            ),
            vis_mode="collision",
        )

    scene.build(n_envs=N_ENVS)

    # Hold the arms posed over the table so they do not collapse (matches the benchmark scene).
    robot.set_dofs_kp(np.full((robot.n_dofs,), 100.0, dtype=np.float32))
    robot.control_dofs_position(torch.zeros((N_ENVS, robot.n_dofs), dtype=gs.tc_float, device=gs.device))

    # Per-env re-drop schedule (host-side countdown; only the small due-index list crosses to the GPU each fire).
    init_state = scene.get_state()
    rng = np.random.default_rng(0)
    countdown = rng.integers(1, RESET_STEPS_MAX + 1, size=N_ENVS) if RESET_STEPS_MAX > 0 else None
    zeros_ndof = torch.zeros((N_ENVS, robot.n_dofs), dtype=gs.tc_float, device=gs.device)

    import quadrants as qd

    print(
        f"genesis={gs.__version__} quadrants={qd.__version__} backend={gs.backend} n_envs={N_ENVS} "
        f"use_hibernation={USE_HIBERNATION} reset_steps_max={RESET_STEPS_MAX}",
        flush=True,
    )

    for i in range(MAX_STEPS):
        if RESET_STEPS_MAX > 0:
            countdown[:] -= 1
            due = np.nonzero(countdown <= 0)[0]
            if due.size:
                envs_idx = torch.as_tensor(due, dtype=gs.tc_int, device=gs.device)
                scene.reset(init_state, envs_idx=envs_idx)  # set_state also wakes hibernated bodies on these envs
                robot.control_dofs_position(zeros_ndof[: due.size], envs_idx=envs_idx)  # re-arm PD hold
                countdown[due] = rng.integers(1, RESET_STEPS_MAX + 1, size=due.size)
        try:
            scene.step()
        except gs.GenesisException as e:
            print(f"\n*** REPRODUCED at step {i + 1}: {e}", flush=True)
            raise
        if (i + 1) % 20 == 0:
            print(f"step {i + 1}/{MAX_STEPS} ok", flush=True)

    print(f"\nNo NaN after {MAX_STEPS} steps -- bug did NOT reproduce this run.", flush=True)


if __name__ == "__main__":
    main()
```

### Steps to Reproduce

If possible, provide a script triggering the bug, e.g.
```python
# code snippet triggering the bug
import genesis as gs
gs.init()
scene = gs.Scene(show_viewer=False)
broken = scene.add_entity(
    gs.morphs.MJCF(file='attachment.xml'),
)
scene.build()
```
Make sure to attach any needed assets (here `attachment.xml`)!

### Expected Behavior

- doesnt crash with nan

### Screenshots/Videos

_No response_

### Relevant log output

```shell

```

### Environment

- OS: [e.g. Ubuntu 24.04, Windows 11 24H2] container
- GPU/CPU [e.g. A100, RTX 4090, M3pr, Intel I9-9900k, Ryzen 5900x] (N/A if no GPU/CPU) rtx 6000 blackwell
- GPU-driver version (N/A if no GPU) 
- CUDA / CUDA-toolkit version (N/A if non-Nvidia)

### Release version or Commit ID

genesis-world main 6a2a1f0c8038c78452cb7ef080cca7f4c38c3a7c. quadrants release assocaited with pyproject.toml.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by duburcqa**
## Description
On GPU, a sleeping body kept receiving constraint forces: the cooperative collision-row assembly had no hibernation branch, so its resting contacts stayed live rows, and the env-wide Newton passes kept stepping its dofs along a stale search direction while the per-island factor skipped its island. Under contact pruning, the wake-on-contact pass also read raw contact slots instead of the kept ones, delaying the wake of a struck sleeper. Both assembly paths now write inert rows through shared helpers, both GPU arms zero the gradient and search direction of hibernated islands, the wake pass and the advection of a sleeper's contacts read contacts through `contact_sort_idx`, a sleeping dof keeps zero velocity and acceleration, and a resting sleeper keeps reporting the support force it is at rest under.

## Related Issue
Resolves https://github.com/Genesis-Embodied-AI/genesis-world/issues/3293

## How Has This Been / Can This Be Tested?
`test_hibernation_wakes_on_collision` now asserts that sleepers keep zero velocity and forces frozen at their weight while an awake body is simulated in the same env, which the previous code fails on GPU, and `test_hibernation_with_pruning` asserts the same weight for sleepers whose contacts are pruned. The script from the issue runs to completion.

## Checklist:
- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [x] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: genesis/engine/solvers/rigid/constraint/solver.py [hunk 1]
```diff
@@ -857,9 +909,27 @@ def _add_collision_constraints_per_friction(
...
-            )
+            is_inert = False
+            if qd.static(rigid_config.use_hibernation):
+                i_col = collider_state.contact_sort_idx[i_col_, i_b]
+                link_a = collider_state.contact_data.link_a[i_col, i_b]
+                link_b = collider_state.contact_data.link_b[i_col, i_b]
+                is_inert = _is_contact_inert(link_a, link_b, i_b, dyn_state, dyn_info, rigid_config)
+            if is_inert:
+                n_con = constraint_state.n_constraints[i_b] + i_col_ * rows_per_contact + i_friction
+                _clear_inert_collision_row(n_con, i_b, constraint_state, rigid_config)
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/solvers/rigid/constraint/solver.py:924]: Real: a sleeper's reported contact force dropped to zero once its rows went inert. An inert contact now keeps the force of the last solve it took part in, and the test asserts that this frozen force is the weight.
